### PR TITLE
feat: Effects adding other effects on removal

### PR DESCRIPTION
**Pull request by Coolthulhu**
## Purpose of change

This PR allows defining comedowns and similar effects that can coexist with the parent effect and can be defined by mods.

## Describe the solution

When effect expires, it can create other effects, at set duration/intensity, or inheriting duration/intensity from the decaying effect.
New effects are added at the end of the effect processing loop, and so should not be able to cause effect map migration. They are added in the same order as the parent effects were processed.
There is an option to only affect decaying (0 duration) effects, only affect removed effects (`remove_effect` when duration > 0), or both.

Reimplemented adrenaline and its comedown in the new system. This means that both effects can now coexist - you can have adrenaline buff with comedown debuff applied at the same time.

## Describe alternatives you've considered

* Waiting for lua to cover it, with associated problems, such as lack of common standard of effect decay
* Leaving it hardcoded, since we don't have many effects that could use this feature (possibly because we didn't have this feature)

## Testing

* Tests pass
* Spawn 3x adrenaline
* Use adrenaline
* Wait for it to decay
* Apply adrenaline
* Confirm both buff and debuff exist
* Wait for both to time out
* Apply adrenaline
* Use blood filtering bionic
* Confirm that adrenaline comedown is present

## Checklist

* [x] Body part inheritance
* [x] Fixing adrenaline:
  * [x] Rebalance duration, to account for debuff no longer being included in "common duration"
  * [x] Remove hardcoded addition of debuff from blood filtering and possibly others
* [x] Docs

### CODE REVIEW

File: src/effect.h [hunk 1]
```diff
@@ -34,6 +35,59 @@ enum effect_rating {
...
+
+struct caused_effect_sort_less {
+    public:
+        bool operator()( const caused_effect &lhs, const caused_effect &rhs ) const {
+            if( lhs.type != rhs.type ) {
+                return lhs.type < rhs.type;
+            }
+            return lhs.intensity_requirement < rhs.intensity_requirement;
+        }
+};
```
Comment (scarf005) [inline src/effect.h:89]:
```suggestion
const auto caused_effect_sort_less = []( const caused_effect &lhs, const caused_effect &rhs ) -> bool {
	return lhs.type == rhs.type ? lhs.intensity_requirement < rhs.intensity_requirement : lhs.type < rhs.type;
};
```

this comparator doesn't seem to be used anywhere in PR, would it be intentional?
also, making it lambda might be better for ergonomics.

**Review by scarf005 — APPROVED** [review]

**Review by scarf005 — APPROVED** [review]

**Review by scarf005 — CHANGES_REQUESTED** [review]
> - Tests pass
> - Spawn 3x adrenaline
> - Use adrenaline
> - Wait for it to decay
> - Apply adrenaline
> - Confirm both buff and debuff exist

image
image
image

works as intended, however on comedown it errors with:
```
 DEBUG    : player_morale::perceived_pain is inconsistent.

 FUNCTION : bool player_morale::consistent_with(const player_morale &) const
 FILE     : /home/scarf/repo/cata/Cataclysm/src/morale.cpp
 LINE     : 878
 VERSION  : BN 8b14530d31ec
```

> - Wait for both to time out
> - Apply adrenaline
> - Use blood filtering bionic
> - Confirm that adrenaline comedown is present

image

even after using blood filter CBM, the effect does not wear off.

File: src/effect.h (hunk 1 above)
Comment (Coolthulhu) [reply src/effect.h:89]: Oh right, I used it for something before, but now it's probably obsolete.

**Comment by Coolthulhu** [comment]
I didn't realize that blood filter was not removing adrenaline. That's why it wasn't triggering the aftereffect.

**Review by scarf005 — APPROVED** [review]
image

LGTM.

**Review by scarf005 — APPROVED** [review]

**Pull request merged**